### PR TITLE
feat: Timeline UI

### PR DESCRIPTION
**Pull request by johnmarksilly**
### Timeline UI

Created a new UI for viewing events and live camera views

https://user-images.githubusercontent.com/11224731/149416284-6a1cba74-b048-4c78-8012-2a05c92d4405.mov

To be completed before PR is ready
- [x] Investigate using the recordings framework for faster loading (load segments instead)
- [x] Add config for using new UI
- [x] Update docs
- [x] Handle no events for today
- [x] Load first timeline entry (most recent) and scroll to that event
- [x] Play/Pause + disable arrows on end indices
- [x] Placeholder sizing while video is loading to reduce "buggy" looking element + missing image
- [x] Move timeline while video plays (timeline will be disabled while video is playing)
- [x] Add global padding back to pages (except new route)
- [x] Make sure change complies with dark mode

[Visual Update 02/15/22]

https://user-images.githubusercontent.com/11224731/154198712-4bae4c47-65c0-47b4-99be-d34081c5ffc8.mov

### CODE REVIEW

**Comment by NickM-27** [comment]
I know it's early but just wanted to say the screenshot looks great!

File: web/src/components/HistoryViewer.jsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import { HistoryHeader } from '../routes/HistoryHeader';
+import { longToDate } from '../utils/dateUtil';
+import Timeline from './Timeline';
+
+export default function HistoryViewer({ camera }) {
+  const apiHost = useApiHost();
+  const videoRef = useRef();
+
+  // const beginningOfDay = new Date().setHour(0, 0, 0) / 1000;
+  const beginningOfDay = new Date(new Date().getTime() - 24 * 60 * 60 * 1000);
```
Comment (johnmarksilly) [inline web/src/components/HistoryViewer.jsx:17]: I'd like this to be day of, and potentially if you reach the end, load the next days? Or would a dropdown be better?

**Comment by NickM-27** [comment]
Looks awesome! Does it also allow viewing parts of the saved recordings that don't have an associated event?

**Comment by johnmarksilly** [comment]
> Looks awesome! Does it also allow viewing parts of the saved recordings that don't have an associated event?

I'm actually working on that, but I think it might be useful to put that in another PR since that's going to take a bit of experimentation. Not everyone uses the recordings so I'd like to figure out a way to integrate it seamlessly without interrupting the experience of those without extra recordings.

I have 24/7 recording so in essence it would replace all the empty space with the recorded content

**Comment by NickM-27** [comment]
> > Looks awesome! Does it also allow viewing parts of the saved recordings that don't have an associated event?
> 
> I'm actually working on that, but I think it might be useful to put that in another PR since that's going to take a bit of experimentation. Not everyone uses the recordings so I'd like to figure out a way to integrate it seamlessly without interrupting the experience of those without extra recordings.
> 
> I have 24/7 recording so in essence it would replace all the empty space with the recorded content
> 
> [EDIT]: This is basically a 1v1 replacement for the old camera page (and moving the last 24 hour events into this page)

Ah I see, that makes sense. I also have 24/7 recording which is why I asked.

**Comment by hcgonzalezpr** [comment]
Looks awesome, almost what I was looking in https://github.com/blakeblackshear/frigate/issues/2230

**Comment by johnmarksilly** [comment]
Sorry for such a long delay! I have been extremely sick and didn't have much time to do anything!

Updates:
- refactored the rendering logic in order to support offsetting the timeline blocks vertically for overlapping events
- refactored to typescript because I couldn't handle not knowing if my objects were consistent. 
- misc fixes

There is an outstanding bug with how the overlapping offset logic works that I have to address

**Comment by johnmarksilly** [comment]
The time has come 🗡️ 

There are some quirks with chrome/android at the moment that I'm still trying to work out. Lots of planned improvements:

- [ ] Loading screens for fetching events
- [ ] Pre-load snapshots that are before and after the marker position for seamless scrubbing
- [ ] Pre-cache timeline events on the backend (similar to Nest) by saving a video for a full day that with 1 frame per minute or something like that
- [ ] Load recording if no events occur at a specified marker time
- [ ] Start video at marker position
- [ ] Don't switch events to an overlapping event while playing an event (this causes the video to load the next event).
- [ ] Refactoring to make things more decoupled. I don't like having to bubble up the play state through the timeline component

Lots more to come. If anyone wants a demo or questions/comments/concerns, post here and maybe I can schedule a Zoom/Discord meeting (I'm in CST)

**Comment by johnmarksilly** [comment]
@blakeblackshear I think this can be re-run. I'm currently deploying to my k8s cluster and the prod build works locally so should be 💯

**Comment by johnmarksilly** [comment]
Eff I forgot to run the tests. I'll fix quick.

**Comment by blakeblackshear** [comment]
Please open a new PR to `release-0.11.0`.

**Comment by johnmarksilly** [comment]
Sounds good. Do you not prefer to change the base branch in open PRs? It's a personal preference, but it maintains the history of the PR/comments if you don't close the branch. I've never had a public repo so I don't know if that's normal to do when having collaborators create PRs against forks.

**Pull request closed without merging**